Add feature to Ironium to test delayed jobs with Timekeeper in test enviroment

ugotsoul opened the pull request:
Issue: https://github.com/assaf/ironium/issues/48

- Currently, there is no easy way to determine when a delayedJob has been processed without moving the system time forward (stubbing system time, or waiting out the duration with `setTimeout`)
- Solution: keep track of the queued time a of a job in a data structure, and kick the job when the queued time + the duration of the delay is less than or the same as the current time (Date.now). Allows the user to test one or more delayed jobs in test environment with Timekeeper. 
- Also: fixes an assertion in delayedJobs where `toString` is not called in the assertion (?)

Commit:
Update delayJob test to use TimeKeeper

Commit:
Add kickJob method to Beanstalkd

Commit:
Revert changes to with_delay_test, change sufficient timeout to 500ms

Commit:
Add methods to runOnce to process delayedJobs in test

Commit:
Handle errors from delayed job not being found

Commit:
Revert change to with_delay_test

Commit:
- add initial tests for delayed job test feature

Commit:
Add tests for multiple delayed jobs in test env

Commit:
Merge remote-tracking branch 'upstream/master' into usa_delayjob2

djanowski commented on lib/queues.js:
```diff
@@ -163,15 +164,24 @@ class Queue {
   // Push job to queue with the given delay.  Delay can be millisecond,
   // or a string of the form '5s', '2m', etc.
   delayJob(job, duration) {
+    const self = this;
```
Some times this is needed, but can be avoided here by using fat-arrow syntax: `.then(jobIDs => {`

djanowski commented on lib/queues.js:
```diff
@@ -163,15 +164,24 @@ class Queue {
+      .then(function(jobIDs) {
+
+        if (process.env.NODE_ENV === 'test')
+          self._delayedJobs.set(jobIDs[0], Date.now() + ms(duration));
```
So this will work only when calling `delayJob()` and not `delayJobs()`?

djanowski replied:
What happens if you pass a duration as a number of milliseconds?

ugotsoul replied:
I thought about moving this logic to delayJobs, but: `delayJobs` isn't an exposed method, and a value for the duration is only passed to `delayJob`. so why go through the trouble of setting a job id/duration in the map for jobs with no duration / no delay. 🤔

ugotsoul replied:
> What happens if you pass a duration as a number of milliseconds?

Thanks! I missed this. It should  handle both a string (`2m`) or milliseconds input for duration.

ugotsoul replied:
I'm going to go with your suggestion - delayJobs already has the assertion against duration's value  + converts duration value for us. 👍

djanowski commented on lib/queues.js:
```diff
@@ -260,13 +270,62 @@ class Queue {
       debug('Waiting for jobs on queue %s', this.name);
-      return this._getClientPromise.then(client => this._reserveAndProcess(client));
+
+      if (process.env.NODE_ENV === 'test' && this._delayedJobs.size > 0) {
```
- `runOnce()` _is_ for testing, so maybe we don't need to check `NODE_ENV` here.
- Anyway, we are checking `NODE_ENV` to populate `_delayedJobs`, so we can trust that if there's anything in there, we want to check for it here.
- `_delayedJobs` is always a collection (empty or not), so maybe we can avoid the `if`?

ugotsoul replied:
What about keeping the if statement for `delayedJobs.size > 0` to avoid going through this logic when there are no delayed jobs?

djanowski commented on lib/queues.js:
```diff
@@ -260,13 +270,62 @@ class Queue {
+            promises.push(promise);
+          }
+
+          return Promise.all(promises);
```
We're requiring Bluebird already, so `Bluebird.map` to the rescue!

djanowski commented on lib/queues.js:
```diff
@@ -260,13 +270,62 @@ class Queue {
+
+      if (process.env.NODE_ENV === 'test' && this._delayedJobs.size > 0) {
+
+        const jobIDs = getReadyJobIDs(this);
```
We're already working with OOP, so maybe we can have "private" methods of the class (`_`-prefixed) instead of passing `this` as an argument?

ugotsoul replied:
That sounds good! I'll move these (this and `kickJobs`) to private methods.

djanowski commented on test/queue/kick_job_test.js:
```diff
@@ -0,0 +1,135 @@
+const TimeKeeper  = require('timekeeper');
+
+
+describe('Queuing a delayed job in test enviroment', function() {
```
Typo in environment.

ugotsoul replied:
Oops. nt gud at spelling

djanowski commented on test/queue/kick_job_test.js:
```diff
@@ -0,0 +1,135 @@
+    });
+
+
+    after(() => processedJobs.splice(0, processedJobs.length));
```
Just `splice()` should have the same effect.
But anyway, isn't the second test (multiple jobs) a good enough test case to cover both 1 or multiple jobs?

ugotsoul replied:
yes! that's a good point. I'll remove the stand alone test, its redundant then.

Commit:
Update tests with PR feedback

- Remove extraneous tests for one job, keep tests for multiple jobs
- Fix spelling error, use setup test helper

Commit:
PR comments

- Move getReadyJobIDs and kickJobs to private methods
- Use Bluebird.map instead of Promise.all in kickJobs
- Store jobID in delayJobs method, not delayJob

ugotsoul commented:
Failing tests because I'm using features from node v6+. Going to remove features to be backwards compatible for node v4 +

Commit:
Use forEach instead of for...of with multiple arguments for node 4 compatiblity

ugotsoul commented:
I'm seeing test failures with `UNKNOWN_COMMAND` errors from beanstalkd. Could beanstalkd version on travis CI be the wrong version? 

Checked the installed version of beanstalkd on travis:
```
4.74s$ sudo apt-get install -qq beanstalkd
Selecting previously unselected package beanstalkd.
(Reading database ... 82140 files and directories currently installed.)
Unpacking beanstalkd (from .../beanstalkd_1.4.6-3_amd64.deb) ...
Processing triggers for ureadahead ...
Processing triggers for man-db ...
Setting up beanstalkd (1.4.6-3) ...
```

The version on the PPA referenced on beanstalkd install documentation is listed as `1.5.0`  https://launchpad.net/~jernej/+archive/ubuntu/beanstalkd

I switched over to the same node version as Travis CI (node 4.8) and removed and reinstalled node packages... test suite ran with no errors or failures. 🤔

djanowski commented:
I updated our Travis configuration to use a more recent Ubuntu image, which gives us Beanstalkd 1.9 (kick command available from 1.8).

djanowski commented on lib/queues.js:
```diff
@@ -163,15 +164,14 @@ class Queue {
   delayJob(job, duration) {
-    return this.delayJobs([ job ], duration)
-      .then(jobIDs => jobIDs[0]);
+    return this.delayJobs([ job ], duration);
```
`delayJob()` now returns an array of job IDs? 🤔

ugotsoul replied:
Oops. Deleted too much. Will add back the `then` call

djanowski commented on lib/queues.js:
```diff
@@ -194,7 +194,17 @@ class Queue {
+          const msDelay = delay * 1000;
+
+          if (process.env.NODE_ENV === 'test' && msDelay > 0) {
+            jobIDs.map(function(id) {
```
`forEach`?

djanowski commented on lib/queues.js:
```diff
@@ -260,16 +270,64 @@ class Queue {
+              if (error.message === 'NOT_FOUND')
+                debug(`Could not find and delete job with id: ${id}`);
+              else
+                throw new Error(`Error deleting job with id: ${id} with message: ${error.message}`);
```
We're kicking, not deleting, but maybe we don't need all this catch block for now?

ugotsoul replied:
Ah. I added this because after kicking a job that was already processed, there were `NOT_FOUND` error in this test: https://github.com/assaf/ironium/blob/master/test/queue/with_delay_test.js#L48

Another solution I looked into: write a peek by id method to check if the job has been processed or not before issuing a kick command.

djanowski replied:
True! Let's leave it as is!

djanowski commented on lib/queues.js:
```diff
@@ -260,16 +270,64 @@ class Queue {
+  _getDelayedJobIDs() {
+    const jobIDs = [];
+
+    this._delayedJobs.forEach(function(readyTime, id) {
```
`Array.from(this._delayedJobs).filter(...)`

ugotsoul replied:
I think using forEach makes the arguments more readable, instead of:

```
    const jobIDs = Array.from(this._delayedJobs)
      .filter((jobData) => jobData[1] <= Date.now());

    return jobIDs;
```
What do you think?

djanowski replied:
Yeah without array destructuring it's hard to say, let's leave `forEach` if you find it better!

djanowski commented on lib/queues.js:
```diff
@@ -260,16 +270,64 @@ class Queue {
 
 
+  // Returns a list of stored delayed jobs IDs that are ready to be processed
+  _getDelayedJobIDs() {
```
So per the comment above the function probably needs the "ready" term in its name?

djanowski commented on lib/queues.js:
```diff
@@ -260,16 +270,64 @@ class Queue {
+
+        const jobIDs = this._getDelayedJobIDs();
+
+        return this._kickJobs(jobIDs)
```
Looks like these functions don't have any overhead apart from looping over the collection, and looping over an empty collection should be a no-op, so why not merge the two branches of the `if`?

ugotsoul replied:
Wouldn't there still be an operation - kickJobs resolves the `_putClientPromise` before looping over the jobIDs collection...?

djanowski replied:
It's an already-resolved promise, almost zero overhead, I think clarity wins?

ugotsoul replied:
D'oh. I'll change this.

djanowski commented:
Should we kick the jobs in the order they would become ready in real life? Right now we're kicking in the order they were queued.

assaf commented:
Why kick jobs to begin with?

djanowski commented:
@assaf You suggest keeping them in memory, without queuing to Beanstalkd?

assaf commented:
No, I'm trying to understand why kicking jobs is necessary?

djanowski commented:
- We queue the jobs with a delay.
- Time traveling happens.
- Beanstalkd not subject to time traveling.
- If Ironium determines that the delayed job is ready by `Date.now()`, it needs to tell Beanstalkd to put the job in the ready queue so everything works as expected.

djanowski commented:
(BTW kick = mark a delayed job as ready)

assaf commented:
So me manage time in two places: in Ironium keeping track of which jobs are ready to run, and in Beanstalkd also keeping track of which jobs are ready to run. But with the first state, is the second state necessary? Or redundant information?

djanowski commented:
How else would we prevent Beanstalkd from returning the job right away? Put a fixed, huge delay?

Unless you're suggesting we allow the job to be reserved, then discard it in Ironium if it's not yet ready, and queue it again?

🤔

assaf commented:
Beanstalkd isn't running any jobs, Ironium is. And it needs to decide which jobs are ready to run, so that state (when to run each job) is necessary. And with that info it can decide which jobs to run, and which not. Reserve and release could do that: reserve all the jobs, release the ones that did not run, pick them again in the next run. Same thing we're doing with concurrency, except one case uses time, and the other uses space.

djanowski commented:
So this test-mode awareness spans across even more parts of the code?

🤔  I'll sleep on it

djanowski commented:
I stand by my containment argument (thanks Assaf for coining it).

With this solution we keep the complexity of test environment in `runOnce()`, which is already a test facility.

And we don't add more complexity to consuming jobs, which is already quite complex.

@assaf What do you say, how should we proceed?

assaf commented:
Containing the test logic complexity inside `runOnce` is the stronger argument, so that wins my vote.

djanowski commented:
Thanks @assaf 🙌

@ugotsoul So, I think we should kick the jobs in the order they would become ready? Right now we're kicking in the order they were queued.

assaf commented:
Is there a non-deterministic order? Like kicking all of them concurrently?

djanowski commented:
That's what we're doing right now. Maybe we should just do that. If the code cares about order, it should time travel / runOnce in steps.

Commit:
PR comments

- No need for if statement in runOnce
- Fat fingers: add back .then block to delayJob
- Change method name: _getDelayedJobIDs to _getReadyDelayedJobIDs

ugotsoul commented:
The failing test in this branch is failing on master too (1/3 build jobs failed):
https://travis-ci.org/assaf/ironium/branches

@djanowski when you are back online lets debug this~
